docs/madr/decisions/056-meshservice-migration.md

```diff
@@ -0,0 +1,165 @@
+
+- On Kubernetes there are two outbound listeners:
+  - Traffic to the Kubernetes DNS name/ClusterIP is load-balanced cross-zone
+  - Traffic to the `<kuma.io/service>.mesh`/VIP is load-balanced cross-zone
+
+##### `MeshService`
+
+- We now have one outbound listener:
+  - Traffic to the generated hostname of the `MeshService` goes through the Cluster IP
+  - This cluster IP is no longer cross-zone load balanced
+
+## Considered Options
+
+* Policies applied to `demo-app_kuma-demo_svc_5000` should also apply to
+  the `MeshService {name: demo-app, namespace: kuma-demo, spec.ports[port == 5000]}`.
+  * we can drop this behavior when we drop `kuma.io/service` support
+* For cross-zone:
+  * don't generate MeshService by default
+  * have a setting to disable using MeshServices
+  * change the default behavior of MeshService from single-zone load balancing to cross-zone.
+    Users would need to manually switch over to local zone load-balancing after having
+    potentially migrated cross-zone services to `MeshMultiZoneService`.
+  * instead of changing the behavior of MeshService, instead also generate
+    MeshMultiZoneService by default, which would reintroduce the cross-zone load balancing behavior.
+
+## Decision Outcome
+
+* Policies applied to `demo-app_kuma-demo_svc_5000` should also apply to
+  the `MeshService `{name: demo-app, namespace: kuma-demo, spec.ports[port == 5000]}`.
+* Change `MeshService` to do cross-zone load balancing by default and have users
+  switch over to local zone plus `MeshMultiZoneServices`
+
+## Pros and Cons of the Options
+
+### Don't generate MeshServices by default
+
+Users would need to enable MeshService generation.
+
+#### Pros
+
+* One of the easiest, most straightforward option as far as stability and UX goes.
+
+#### Cons
+
+* Will be harder to motivate users to switch at all
+* None of the advantages of MeshService & HostnameGenerator "for free"
+* MeshServices need to be deleted if generation is enabled then disabled
+
+### Generate but don't enable by default
+
+Users would need to enable usage of MeshService resources.
+
+#### Pros
+
+* One of the easiest, most straightforward option as far as stability and UX goes.
+
+#### Cons
+
+* Will be harder to motivate users to switch at all
+* None of the advantages of MeshService & HostnameGenerator "for free"
+* Odd that MeshService objects exist but aren't used
+
+### Make cross-zone the default behavior
+
+MeshServices by default load balance cross-zone, like `kuma.io/service`. Users
+would disable this behavior to switch to local zone behavior.
+
+#### Pros
+
+* Users get MeshService, HostnameGenerator advantages without breaking the data
+  plane
+
+#### Cons
+
+* Is it coherent?
+  - the set of endpoints is shared between MeshServices local
+    and synced from other zones
+  - HostnameGenerator can't/shouldn't by default generate with a zone suffix
+* Need to keep using ZoneIngress availableServices
+
+### Generate MeshMultiZoneService
```

Comment:
There are a couple of open questions here.

1) How would we preserve policy matching in this case?

policy matching
```
kind: MeshService
name: redis_kuma-demo_svc_5678
```
would not match MeshMultiZoneService, right?

2) How would we switch to using MeshMultiZoneService outbound?
so let say someone is using ClusterIP, we now have to switch so it's using MeshMultiZoneService instead of kuma.io/service. How do we then go to the desired behavior that uses MeshService?

Reply:
1.
> would not match MeshMultiZoneService, right?

I mean it could, right?

2. You mean how to go from cross-zone to local zone behavior and switch to MeshService right? I could imagine:
  - Disable autogeneration
  - Delete MeshMultiZoneServices for services they don't want to do cross-zone

It does feel trickier than a switch to change the behavior for one resource.

Reply:
Yeah. I'm asking those questions to lay out complexity. On paper, it sounds best to "just generate" MeshMultiZoneService, but it's a lot of moving parts.

Reply:
Yeah I don't disagree and probably wouldn't go with this option at this point